docs/source/customizations/native.mdx

```diff
@@ -5,7 +5,183 @@ description: Building and compiling a custom plugin for the Apollo Router with R
 
 import { Link } from 'gatsby';
 
-TODO:
+The router manages a pipeline for GraphQL data. That pipeline takes client Requests and returns Responses. Requests are processed in a structured fashion, each stage taking input Requests and then passing the (potentially modified) Request onto the next stage. After the final stage finishes processing Responses are returned and the pipeline is reversed with each stage now passing a (potentially modified) Response back to the client.
 
-- How to get started with Apollo Router Rust plugins (hello world)
+Router operators may require additional behaviour which is not part of the "standard" pipeline processing: for instance transforming headers, verifying JWT tokens, granting roles, augment responses, etc... It's clearly not viable to encode all of these possibilities within the standard control flow, since not only is it diffcult to predict which additional behaviour is required, but that additional behaviour may not be required in all deployments. The plugin mechanism satisfies this requirement for optional additional processing. Plugins may modify Requests/Responses and also may perform additional tasks at different pipeline stages as required.
+
+### Designing your plugin
+
+The overview describes the plugin pipeline and the first decision to be made when designing your plugin is which of the four stages to implement. By default each of the plugin stages simply passes the request/response to the next service, so it's fine if your plugin doesn't implement a stage. There are four options:
+
+#### Router (router_service)
+
+Implement this function if your plugin must interact at the earliest/latest stage of the pipeline. For example, this would be a good place to perform tasks such as JWT verification before allowing a Request to proceed further.
+
+#### Query Planning (query_planning_service)
+
+Implement this function if your plugin must interact with the query planning functionality. For example,  ???
```

Comment:
While we don't allow our users to mutate a query plan, they could dump the query plan / log it somewhere